Added mmap for StorageFile

pufit opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/run_check.py
-->
### Changelog category (leave one):
- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added mmap support for StorageFile, which should improve the performance of clickhouse-local


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/


#### Related Issue
https://github.com/ClickHouse/ClickHouse/issues/5149

#### Benchmarks
[./run.sh](https://github.com/ClickHouse/ClickBench/blob/main/clickhouse-local/run.sh) results

<details>
  <summary>Old version</summary>
  
  ```
[5.685, 0.071, 0.081],
[5.564, 0.087, 0.068],
[9.332, 0.123, 0.121],
[9.021, 0.143, 0.133],
[9.122, 0.389, 0.497],
[11.050, 0.618, 0.612],
[5.667, 0.078, 0.066],
[5.682, 0.079, 0.079],
[10.249, 0.566, 0.577],
[19.155, 0.648, 0.658],
[14.395, 0.343, 0.334],
[14.474, 0.365, 0.362],
[11.079, 0.615, 0.591],
[19.813, 0.945, 0.899],
[11.499, 0.676, 0.704],
[9.213, 0.518, 0.526],
[20.108, 1.534, 1.514],
[19.672, 1.103, 1.089],
[31.692, 2.953, 2.950],
[8.939, 0.147, 0.145],
[49.022, 1.893, 1.893],
[59.560, 2.191, 2.072],
[108.502, 3.724, 3.742],
[238.224, 231.524, 228.034],
[22.236, 0.489, 0.491],
[10.941, 0.364, 0.385],
[22.204, 0.486, 0.498],
[53.378, 2.210, 2.129],
[43.669, 9.491, 9.492],
[5.700, 1.178, 1.184],
[22.825, 0.666, 0.644],
[43.234, 0.902, 0.871],
[35.278, 3.585, 3.562],
[49.831, 2.937, 2.970],
[49.852, 3.007, 2.966],
[7.456, 0.969, 0.967],
[63.925, 1.910, 1.865],
[59.929, 1.829, 1.908],
[63.547, 1.855, 1.860],
[96.735, 2.796, 2.809],
[30.907, 0.402, 0.393],
[31.007, 0.385, 0.373],
[22.203, 0.334, 0.315],
```
</details>

<details>
  <summary>mmap version</summary>
  
```
[1.457, 0.083, 0.080],
[1.086, 0.071, 0.067],
[3.076, 0.114, 0.099],
[6.427, 0.129, 0.145],
[6.560, 0.378, 0.362],
[8.304, 0.603, 0.607],
[1.109, 0.080, 0.066],
[1.116, 0.077, 0.076],
[7.792, 0.545, 0.555],
[10.643, 0.683, 0.645],
[7.418, 0.328, 0.326],
[7.782, 0.376, 0.363],
[8.372, 0.609, 0.602],
[14.509, 0.861, 0.821],
[8.744, 0.684, 0.708],
[6.639, 0.640, 0.524],
[14.853, 1.508, 1.527],
[14.431, 1.102, 1.102],
[24.058, 2.938, 2.927],
[6.437, 0.125, 0.142],
[47.368, 1.948, 1.892],
[55.282, 2.165, 2.081],
[103.264, 3.791, 4.017],
[255.104, 241.867, 217.470],
[16.833, 0.488, 0.495],
[8.240, 0.372, 0.369],
[16.818, 0.521, 0.460],
[48.315, 2.285, 2.177],
[41.492, 9.461, 9.426],
[2.427, 1.247, 1.163],
[16.061, 0.652, 0.655],
[32.406, 0.872, 0.890],
[26.197, 3.564, 3.556],
[48.199, 3.139, 3.170],
[48.107, 3.030, 2.981],
[5.288, 0.945, 0.965],
[49.856, 2.203, 1.836],
[45.572, 1.939, 1.838],
[49.667, 2.017, 1.842],
[89.379, 3.031, 3.016],
[18.126, 0.387, 0.370],
[15.024, 0.354, 0.360],
[10.761, 0.310, 0.298],
```
</details>

#### Comment
I've run benchmarks on my devices, so the result differs from https://benchmark.clickhouse.com/. 
Still, we can see a slight improvement in the cold run almost in all queries. 
<details>

![image](https://user-images.githubusercontent.com/19871338/205524888-10b8dd97-e932-4ccd-9f42-843f719cecba.png)
</details>

However, there is no improvement in the hot run.
<details>

![image](https://user-images.githubusercontent.com/19871338/205524983-50cefa35-e1f7-4649-ba0e-c779fb9c8489.png)
</details>

pufit committed:
Added mmap for StorageFile

pufit committed:
Merge branch 'master' into mmap-for-storage-file

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=43927) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=43927) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=43927) it.</sub>

pufit committed:
Use mmap only on regular files.

pufit committed:
Merge remote-tracking branch 'origin/mmap-for-storage-file' into mmap-for-storage-file

pufit commented:
Small note: it seems that `MMapReadBufferFromFile` actually doesn't throw any exceptions, even if the input file couldn't be mmapped. And on the other hand, I found this part of [code](https://github.com/ClickHouse/ClickHouse/blob/86a0dd010f0a95af1991a363f60c57fd1737f37b/src/Disks/IO/createReadBufferFromFileBase.cpp#L61), which relies on an exception been thrown in case a file descriptor was a pipe.
Is that a bug?

pufit committed:
Merge branch 'master' into mmap-for-storage-file

pufit commented on src/Storages/StorageFile.cpp:
```diff
@@ -208,7 +210,7 @@ std::unique_ptr<ReadBuffer> createReadBuffer(
 
         if (S_ISREG(file_stat.st_mode))
-            nested_buffer = std::make_unique<ReadBufferFromFilePRead>(current_path, context->getSettingsRef().max_read_buffer_size);
+            nested_buffer = std::make_unique<MMapReadBufferFromFile>(current_path, 0);
```
This currently doesn't respect the `max_read_buffer_size` settings, leading to several test failures. Will think about how to fix it tomorrow.

alexey-milovidov replied:
It is ok - the setting `max_read_buffer_size` is irrelevant for the case when `mmap` is used for reading.

How to adjust the tests:
- introduce a setting `storage_file_allow_mmap` with value 1 by default and set it to 0 in those tests.

It makes sense to introduce this setting nevertheless as a feature flag (to allow users to turn the new behavior on and off without downgrading in case of any troubles happen).

pufit replied:
In the end, I've decided completely move all this logic to already existing functionality in `Disks/IO/createReadBufferFromFileBase`. I also thought to add a new setting but found the `local_filesystem_read_method` settings, which is basically doing what we are looking for. It also makes sense that `StorageFile` itself should respect `IO/ReadSettings`.

nikitamikhaylov commented:
A test is required. There are two options: launch clickhouse-local under strace and count mmap's. Or use some metric from profile events from system.query_log

alexey-milovidov commented:
> Small note: it seems that MMapReadBufferFromFile actually doesn't throw any exceptions, even if the input file couldn't be mmapped.

That's strange - I see an exception should be thrown from `MMappedFileDescriptor::set`.

pufit committed:
Rewrite `StorageFile` buffer creation with `createReadBufferFromFileBase`.
Add file descriptor support for `createReadBufferFromFileBase`.
Fix file_size overflow in `createReadBufferFromFileBase`.
Fix `MMapReadBufferFromFileWithCache` file_size definition.

pufit commented:
> > Small note: it seems that MMapReadBufferFromFile actually doesn't throw any exceptions, even if the input file couldn't be mmapped.
> 
> That's strange - I see an exception should be thrown from `MMappedFileDescriptor::set`.

I've added a S_ISREG check, which kinda makes sense since we don't want to mmap any other files. Also, there was a size_t overflow `file_size.value_or(-1)`, so probably this function wasn't test covered properly.

pufit committed:
Try to fix codestyle error.

pufit committed:
Test and codestyle fix.

alexey-milovidov commented:
We want to enable mmap by default for StorageFile, but not for other tables, so we need another setting.
What about: `storage_file_read_method`?

pufit committed:
Add `storage_file_read_method` setting.

pufit committed:
Fix UB, fix test.

pufit committed:
Move reader selection logic back to `StorageFile`.

pufit committed:
dummy fix, additional test

pufit committed:
ya fix.

pufit committed:
Merge branch 'master' into mmap-for-storage-file

Avogar commented on src/Disks/IO/createReadBufferFromFileBase.cpp:
```diff
@@ -52,7 +52,12 @@ std::unique_ptr<ReadBufferFromFileBase> createReadBufferFromFileBase(
         {
-            auto res = std::make_unique<MMapReadBufferFromFileWithCache>(*settings.mmap_cache, filename, 0, file_size.value_or(-1));
+            std::unique_ptr<MMapReadBufferFromFileWithCache> res;
+            if (file_size)
+                res = std::make_unique<MMapReadBufferFromFileWithCache>(*settings.mmap_cache, filename, 0, *file_size);
+            else
+                res = std::make_unique<MMapReadBufferFromFileWithCache>(*settings.mmap_cache, filename, 0, *file_size);
```
There are two identical lines in both branches

Avogar commented on src/Core/Settings.h:
```diff
@@ -592,6 +592,7 @@ static constexpr UInt64 operator""_GiB(unsigned long long value)
     M(UInt64, function_range_max_elements_in_block, 500000000, "Maximum number of values generated by function 'range' per block of data (sum of array sizes for every row in a block, see also 'max_block_size' and 'min_insert_block_size_rows'). It is a safety threshold.", 0) \
     M(ShortCircuitFunctionEvaluation, short_circuit_function_evaluation, ShortCircuitFunctionEvaluation::ENABLE, "Setting for short-circuit function evaluation configuration. Possible values: 'enable' - use short-circuit function evaluation for functions that are suitable for it, 'disable' - disable short-circuit function evaluation, 'force_enable' - use short-circuit function evaluation for all  ...
     \
+    M(String, storage_file_read_method, "mmap", "Method of reading data from storage file, one of: read, pread, mmap, pread_threadpool.", 0) \
```
It would be better to use enum here instead of String, so this value will be checked when we set it, because otherwise we can set some random value and get an exception only when using engine `file`. See file `SettingsEnum.h`. We can reuse `LocalFSReadMethod` there and declare only 3 possible values for this setting `mmap/pread/read`.

pufit committed:
Enum settings, fix else branch.

pufit committed:
Merge remote-tracking branch 'origin/mmap-for-storage-file' into mmap-for-storage-file

pufit committed:
Fix codestyle, fix test.

pufit committed:
Fix test codestyle.

Avogar committed:
Fix check black

Avogar committed:
Try fix build

Avogar commented:
Some strange build errors in some build tests. I will checkout this branch locally and try to fix it. I hope I can merge it soon.

Avogar committed:
Merge branch 'master' into mmap-for-storage-file

Avogar committed:
Fix builds

Avogar commented:
`Stress test (asan) — Sanitizer assert (in stderr.log)` - https://github.com/ClickHouse/ClickHouse/issues/45199
